sycl/include/sycl/ext/oneapi/experimental/named_sub_group_sizes.hpp

```diff
@@ -0,0 +1,45 @@
+//== named_sub_group_sizes.hpp --- SYCL extension for named sub-group sizes ==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/oneapi/kernel_properties/properties.hpp>
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext::oneapi::experimental {
+
+struct named_sub_group_size {
+  static constexpr uint32_t primary = 0;
+  static constexpr uint32_t automatic = -1;
+};
+
+inline constexpr sub_group_size_key::value_t<named_sub_group_size::primary>
+    sub_group_size_primary;
+
+inline constexpr sub_group_size_key::value_t<named_sub_group_size::automatic>
+    sub_group_size_automatic;
+
+namespace detail {
+template <>
+struct PropertyMetaInfo<
+    sub_group_size_key::value_t<named_sub_group_size::automatic>> {
+  // sub_group_size_automatic means that the kernel can be compiled with
+  // any sub-group size. That is, if the kernel has the sub_group_size_automatic
+  // property, then no sycl-sub-group-size IR attribute needs to be attached.
+  // Specializing PropertyMetaInfo for sub_group_size_automatic and setting
+  // name to an empty string will result in no sycl-sub-group-size IR being
+  // attached.
+  static constexpr const char *name = "";
+  static constexpr const char *value = 0;
+};
```

Comment:
I think this is correct? At least according to my intention/design: since the named sub group size properties are just special values of the `sub_group_size` property, without any specialization, launching a kernel with `sub_group_size_automatic` will attach `sycl-sub-group-size` with value `named_sub_group_sizes::automatic=-1` and `sub_group_size_primary` will attach `sycl-sub-group-size` with value `named_sub_group_sizes::primary=0`. Then,

- for `automatic`, we don't want this behavior, so we speciaize this and set the name to the empty string to avoid attaching `sycl-sub-group-size`;
- for `primary`, we keep this behavior, which will means that `sycl-sub-group-size = 0` metadata will be attached to the kernel. We then use the presence of the `sycl-sub-group-size = 0` metadata to signal the use of the `sub_group_size_primary`.

The test I added is also passing, although that only checks for the presence/non-presence of `intel_reqd_sub_group_size`, which is what the frontend translates `sycl-sub-group-size` into. I do admit the design of using special values in `intel_reqd_sub_group_size`/`sycl-sub-group-size` to signal the use of a named sub group size is a bit awkward, and I wouldn't be opposed to changing it, but that would required some small additions in `sycl-post-link` to get module splitting to work.

Reply:
Apologies, I now see what you are up to here. Clever! I think this solution is good. 👍